lib/protobuf/json.ex

```diff
@@ -0,0 +1,161 @@
+  | `Wrapper types`          | `various types`       | No        |
+  | `FieldMask`              | `string`              | No        |
+  | `ListValue`              | `array [foo, bar, …]` | No        |
+  | `Value`                  | `value`               | No        |
+  | `NullValue`              | `null`                | No        |
+  | `Empty`                  | `object`              | No        |
+
+  ## Usage
+
+  `Protobuf.JSON` requires a JSON library to work, so first make sure you have `:jason` added
+  to your dependencies:
+
+      defp deps do
+        [
+          {:jason, "~> 1.2"},
+          # ...
+        ]
+      end
+
+  With `encode/1` you can turn any `Protobuf` message struct into a JSON string:
+
+      iex> message = Car.new(color: :RED, top_speed: 125.3)
+      iex> Protobuf.JSON.encode(message)
+      {:ok, "{\\"color\\":\\"RED\\",\\"topSpeed\\":125.3}"}
+
+  JSON keys are encoded as camelCase strings, specified by the `json_name` field options. So
+  make sure to recompile the `.proto` files in your project before working with JSON encoding.
+  You can set your own `json_name` too:
+
+      message GeoCoordinate {
+        double latitude = 1 [ json_name = "lat" ];
+        double longitude = 2 [ json_name = "long" ];
+      }
+
+  """
+
+  alias Protobuf.JSON.{Encode, EncodeError}
+
+  @type encode_opt :: {:json_library, module} | Encode.encode_opt()
+
+  @doc """
+  Generates a JSON representation of the given protobuf `struct`.
+
+  Similar to `encode/1` except it will unwrap the error tuple and raise in case of errors.
+
+  ## Examples
+
+      iex> Car.new(top_speed: 80.0) |> Protobuf.JSON.encode!()
+      ~S|{"topSpeed":80.0}|
+
+      iex> TestMsg.Foo2.new() |> Protobuf.JSON.encode!()
+      ** (Protobuf.JSON.EncodeError) JSON encoding of 'proto2' syntax is unsupported, try proto3
+  """
+  @spec encode!(struct, [encode_opt]) :: String.t() | no_return
+  def encode!(%_{} = struct, opts \\ []) do
+    case encode(struct, opts) do
+      {:ok, json} -> json
+      {:error, error} -> raise error
+    end
+  end
+
+  @doc """
+  Generates a JSON representation of the given protobuf `struct`.
+
+  ## Options
+
+    * `:use_proto_names` - use original field `name` instead of the camelCase `json_name` for
+      JSON keys (default is `false`).
+    * `:use_enum_numbers` - encode `enum` field values as numbers instead of their labels
+      (default is `false`).
+    * `:emit_unpopulated` - emit all fields, even when they are blank, empty or set to their
+      default value (default is `false`).
+
+  ## Examples
+
+  Suppose these are you proto modules:
+
+      defmodule Color do
+        use Protobuf, syntax: :proto3, enum: true
+        field :GREEN, 0
+        field :RED, 1
+      end
+
+      defmodule Car do
+        use Protobuf, syntax: :proto3
+        field :color, 1, type: Color, enum: true
+        field :top_speed, 2, type: :float, json_name: "topSpeed"
+      end
```

Comment:
This is a minor nitpick but I _think_ it would be nicer to show these as Protobuf schemas, and not as their compiled Elixir output (since you usually never touch that by hand)?

Reply:
Excellent idea! 💡